# **AYUDANTÍA 4 - Introducción a la Ciencia de Datos - IMT2200**
- Rocío Ladrón de Guevara (rociohgj@uc.cl)
- Daniela Victorero (dvictorerol@uc.cl)


In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# MANEJO DE STRINGS

Cuando trabajamos con datos, esta información nos puede llegar en diferentes formatos, por ejemplo, si guardamos la informacion de nombres de usuarios los guardamos como Strings.

Imaginemos que creamos una pagina web, donde la persona ingresa su nombre, pero se nos olvida indicarle que formato debe estar escrito (si todo en minusculas, mayúsculas o una combinación en eso)

In [ ]:
u = 'AMELIA LINDA'
l = 'amelia linda'
ul = 'AmelIa LinDa'



Nos podria llegar en estos tres formatos.

Consultemos que formato tienen los strings

In [ ]:
print(f'{u} es miniscula {u.islower()}')
print(f'{u} es mayuscula {u.isupper()}')

print(f'{l} es miniscula {l.islower()}')
print(f'{l} es mayuscula {l.isupper()}')

print(f'{ul} es miniscula {ul.islower()}')
print(f'{ul} es mayuscula {ul.isupper()}')


AMELIA LINDA es miniscula False
AMELIA LINDA es mayuscula True
amelia linda es miniscula True
amelia linda es mayuscula False
AmelIa LinDa es miniscula False
AmelIa LinDa es mayuscula False


¿Cómo podemos cambiar de formato a cada uno de los strings?

In [ ]:
#Queremos que sea todo en minúsculas
u.lower()
ul.lower()

#Queremos que sea todo en mayúscula
l.upper()
ul.upper()

#Queremos que el caracter inicial sea mayúscula

u.capitalize()

#Formato título

u.title()

'Amelia Linda'

¿Y que tipo de caracter puede tener los strings?
A veces nuestros strings pueden estar compuetos por digitos, simbolos o caracter alfanumérico.

Python tiene metodos para saber identificar la composición de los strings.

In [ ]:
alp='a13rghhh'

In [ ]:
#Es u  compuesto por caracteres del alfabeto
u.isalpha()


#Es alp alfanumerico

alp.isalnum()

#Es u numerico

u.isnumeric()


False

Y hay más comandos de este estilo.

## Modifiquemos

Alguien me entrego un mensaje, ¡Pero tiene unos caracteres sobrantes!

¡Veamos si podemos formatearlo!

Metodo Strip()

In [1]:
mensaje='{}[][Felices]Fiestas${Patrias}'
mensaje = mensaje.strip('{}[]$')
mensaje


Felices]Fiestas${Patrias


Elimina los caracteres que entregamos como argumento que esten tanto al inicio o al final del strigs

Metodo split()

In [ ]:
mensaje = mensaje.split(']',1)
mensaje


Separa un string en una lista con el separador que indiquemos en el argumento.


In [ ]:
mensaje = ' '.join(mensaje)
mensaje

In [ ]:
mensaje = ' '.join(mensaje.split('${'))
mensaje

Felices Fiestas Patrias


# LIMPIEZA DE DATOS 🧹 💽

Es el proceso de identificar aquella
parte de la data incorrecta,
incompleta, imprecisa, irrelevante o
faltante, y luego modificar,
reemplazar o eliminar según
corresponda

*Fuente: Clase 10 - Introducción a la Ciencia de Datos.*

# DATASET: IMDB - TOP PELÍCULAS Y SERIES DE NETFLIX 🎥

*Fuente: https://www.kaggle.com/datasets/bharatnatrayn/movies-dataset-for-feature-extracion-prediction/code?datasetId=1512088&sortBy=dateRun&tab=profile*

*Información de las columnas*:

- *Movies* : Nombre de las películas o series

- *Year* : Año de estreno o años en los que se transmiten

- *Genre* : Genero(s) de las películas o series.

- *Rating* : Medida de puntuación de las películas o series  de acuerdo a los usuarios.

- *One-Line* : Sinopsis de las películas o series.

- *Stars* : Actores y/o actrices de las películas o series.

- *Votes* : Cantidad de votos de las películas o series por los usuarios.

- *RunTime* : Duración de las películas o series.

- *Gross* : Cantidad de dinero recaudado por la película o serie (a nivel mundial).


In [ ]:
df_movies = pd.read_csv('movies.csv')
df_movies.head(10)

MOVIES         YEAR  \
0                        Blood Red Sky       (2021)   
1  Masters of the Universe: Revelation     (2021– )   
2                     The Walking Dead  (2010–2022)   
3                       Rick and Morty     (2013– )   
4                      Army of Thieves       (2021)   
5                          Outer Banks     (2020– )   
6      The Last Letter from Your Lover       (2021)   
7                               Dexter  (2006–2013)   
8                    Never Have I Ever     (2020– )   
9                         Virgin River     (2019– )   

                                        GENRE  RATING  \
0      \nAction, Horror, Thriller                 6.1   
1  \nAnimation, Action, Adventure                 5.0   
2       \nDrama, Horror, Thriller                 8.2   
3  \nAnimation, Adventure, Comedy                 9.2   
4         \nAction, Crime, Horror                 NaN   
5          \nAction, Crime, Drama                 7.6   
6                \nDrama, Romance                 6.8   
7         \nCrime, Drama, Mystery                 8.6   
8                        \nComedy                 7.9   
9                \nDrama, Romance                 7.4   

                                            ONE-LINE  \
0  \nA woman with a mysterious illness is forced ...   
1  \nThe war for Eternia begins again in what may...   
2  \nSheriff Deputy Rick Grimes wakes up from a c...   
3  \nAn animated series that follows the exploits...   
4  \nA prequel, set before the events of Army of ...   
5  \nA group of teenagers from the wrong side of ...   
6  \nA pair of interwoven stories set in the past...   
7  \nBy day, mild-mannered Dexter is a blood-spat...   
8  \nThe complicated life of a modern-day first g...   
9  \nSeeking a fresh start, nurse practitioner Me...   

                                               STARS    VOTES  RunTime Gross  
0  \n    Director:\nPeter Thorwarth\n| \n    Star...   21,062    121.0   NaN  
1  \n            \n    Stars:\nChris Wood, \nSara...   17,870     25.0   NaN  
2  \n            \n    Stars:\nAndrew Lincoln, \n...  885,805     44.0   NaN  
3  \n            \n    Stars:\nJustin Roiland, \n...  414,849     23.0   NaN  
4  \n    Director:\nMatthias Schweighöfer\n| \n  ...      NaN      NaN   NaN  
5  \n            \n    Stars:\nChase Stokes, \nMa...   25,858     50.0   NaN  
6  \n    Director:\nAugustine Frizzell\n| \n    S...    5,283    110.0   NaN  
7  \n            \n    Stars:\nMichael C. Hall, \...  665,387     53.0   NaN  
8  \n            \n    Stars:\nMaitreyi Ramakrish...   34,530     30.0   NaN  
9  \n            \n    Stars:\nAlexandra Breckenr...   27,279     44.0   NaN

Una sola mirada a los 10 primeras filas del dataframe nos permite realizar varias observaciones:

- Las columnas *GENRE*, *ONE-LINE* y *STARS* contienen saltos de línea (\n). No son necesarios, pues ya hemos procesado los datos en un dataframe.

- Existen datos faltantes en algunas celdas. Hay que tomar una desición de que hacer en estos casos de acuerdo al tipo de dato, información que tenemos, y objetivo de nuestro análisis.

- La columna *Gross* tiene muchos datos faltantes. Comprobaremos que mayoritariamente se compone de estos en los siguientes pasos.

- Las columnas *RunTime* y *Gross* están escritas de forma capitalizada, mientras que el resto de las columnas tienen sus nombres completamente en mayúsculas. Es importante mantener un formato para que a la hora de trabajar con datos.

Dedicarnos a sólo observar una parte del DataFrame no es suficiente, por lo que debemos seguir explorando los datos antes de comenzar a limpiar:

In [ ]:
filas = len(df_movies)

f"Hay {filas} filas en el dataframe df_movies"

'Hay 9999 filas en el dataframe df_movies'

In [ ]:
df_movies.columns

Index(['MOVIES', 'YEAR', 'GENRE', 'RATING', 'ONE-LINE', 'STARS', 'VOTES',
       'RunTime', 'Gross'],
      dtype='object')

In [ ]:
df_movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9999 entries, 0 to 9998
Data columns (total 9 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   MOVIES    9999 non-null   object 
 1   YEAR      9355 non-null   object 
 2   GENRE     9919 non-null   object 
 3   RATING    8179 non-null   float64
 4   ONE-LINE  9999 non-null   object 
 5   STARS     9999 non-null   object 
 6   VOTES     8179 non-null   object 
 7   RunTime   7041 non-null   float64
 8   Gross     460 non-null    object 
dtypes: float64(2), object(7)
memory usage: 703.2+ KB


De acuerdo a la información, podemos hacer ciertas observaciones:

- En el dataframe observamos que la columna *VOTES* corresponden a números enteros, sin embargo, tiene un dtype *object* (correspondiente a un dato categórico).

- De 9999 filas de datos, en la columna *GROSS* sólo 460 datos son no-nulos ¿Que tanto finalmente nos aporta esta información a un posterior análisis?

In [ ]:
df_movies.isnull().head() #podemos obtener un dataframe que nos indica con valores de verdad si una celda es Null

MOVIES   YEAR  GENRE  RATING  ONE-LINE  STARS  VOTES  RunTime  Gross
0   False  False  False   False     False  False  False    False   True
1   False  False  False   False     False  False  False    False   True
2   False  False  False   False     False  False  False    False   True
3   False  False  False   False     False  False  False    False   True
4   False  False  False    True     False  False   True     True   True

In [ ]:
df_movies[df_movies['YEAR'].isnull()].head()  #comprobamos cuales filas tienen datos nulos en 'YEAR', entregando la condición.

MOVIES YEAR  \
878            Vikings: Valhalla  NaN   
1375               Resident Evil  NaN   
1414                    Bright 2  NaN   
1419  Avatar: The Last Airbender  NaN   
1425                   Home Team  NaN   

                                         GENRE  RATING  \
878     \nAction, Adventure, Drama                 NaN   
1375      \nAction, Horror, Sci-Fi                 NaN   
1414    \nAction, Adventure, Crime                 NaN   
1419  \nAction, Adventure, Fantasy                 NaN   
1425               \nComedy, Sport                 NaN   

                                               ONE-LINE  \
878   \nFollow-up series to 'Vikings' set 100 years ...   
1375  \nNearly three decades after the discovery of ...   
1414  \nPlot unknown. Sequel to the 2017 film 'Bright.'   
1419  \nLive-action adaptation of the animated serie...   
1425  \nThe story about New Orleans Saints head coac...   

                                                  STARS VOTES  RunTime Gross  
878   \n            \n    Stars:\nLaura Berlin, \nSa...   NaN      NaN   NaN  
1375  \n            \n    Stars:\nLance Reddick, \nS...   NaN      NaN   NaN  
1414  \n            \n    Stars:\nWill Smith, \nJoel...   NaN      NaN   NaN  
1419                                                 \n   NaN      NaN   NaN  
1425  \n    Directors:\nCharles Francis Kinnane, \nD...   NaN      NaN   NaN

In [ ]:
df_movies[df_movies['MOVIES'].isnull()].head() #no hay valores nulos en la columna MOVIES

Empty DataFrame
Columns: [MOVIES, YEAR, GENRE, RATING, ONE-LINE, STARS, VOTES, RunTime, Gross]
Index: []

In [ ]:
df_movies[df_movies.duplicated()] #comprobamos cuales son estas filas, entregando la condición de si una fila es repetida o no

MOVIES      YEAR  \
6833   Mighty Little Bheem  (2019– )   
6835   Mighty Little Bheem  (2019– )   
6836   Mighty Little Bheem  (2019– )   
6837   Mighty Little Bheem  (2019– )   
6838   Mighty Little Bheem  (2019– )   
...                    ...       ...   
9989                  1899  (2022– )   
9990                  1899  (2022– )   
9991                  1899  (2022– )   
9992                  1899  (2022– )   
9994        The Imperfects  (2021– )   

                                          GENRE  RATING        ONE-LINE  \
6833  \nAnimation, Short, Adventure                 NaN  \nAdd a Plot\n   
6835  \nAnimation, Short, Adventure                 9.0  \nAdd a Plot\n   
6836  \nAnimation, Short, Adventure                 9.0  \nAdd a Plot\n   
6837  \nAnimation, Short, Adventure                 NaN  \nAdd a Plot\n   
6838  \nAnimation, Short, Adventure                 NaN  \nAdd a Plot\n   
...                                         ...     ...             ...   
9989       \nDrama, History, Horror                 NaN  \nAdd a Plot\n   
9990       \nDrama, History, Horror                 NaN  \nAdd a Plot\n   
9991       \nDrama, History, Horror                 NaN  \nAdd a Plot\n   
9992       \nDrama, History, Horror                 NaN  \nAdd a Plot\n   
9994    \nAdventure, Drama, Fantasy                 NaN  \nAdd a Plot\n   

                                                  STARS VOTES  RunTime Gross  
6833  \n    Directors:\nRajiv Chilaka, \nKrishna Moh...   NaN      NaN   NaN  
6835  \n    Directors:\nRajiv Chilaka, \nKrishna Moh...     6      NaN   NaN  
6836  \n    Directors:\nRajiv Chilaka, \nKrishna Moh...     6      NaN   NaN  
6837  \n    Directors:\nRajiv Chilaka, \nKrishna Moh...   NaN      NaN   NaN  
6838  \n    Directors:\nRajiv Chilaka, \nKrishna Moh...   NaN      NaN   NaN  
...                                                 ...   ...      ...   ...  
9989                   \n    Director:\nBaran bo Odar\n   NaN      NaN   NaN  
9990                   \n    Director:\nBaran bo Odar\n   NaN      NaN   NaN  
9991                   \n    Director:\nBaran bo Odar\n   NaN      NaN   NaN  
9992                   \n    Director:\nBaran bo Odar\n   NaN      NaN   NaN  
9994  \n            \n    Stars:\nMorgan Taylor Camp...   NaN      NaN   NaN  

[431 rows x 9 columns]

Tenemos 431 filas con distintas películas y/o series repetidas (compartiendo exactamente los mismos valores en las columnas posteriores)...


In [ ]:
df_movies[df_movies['MOVIES'].duplicated()]

MOVIES         YEAR  \
309            Snowpiercer       (2013)   
435                Kingdom  (2014–2017)   
576   Beauty and the Beast       (1991)   
589                Revenge  (II) (2017)   
990                   Safe   (I) (2012)   
...                    ...          ...   
9991                  1899     (2022– )   
9992                  1899     (2022– )   
9994        The Imperfects     (2021– )   
9997        The Imperfects     (2021– )   
9998        The Imperfects     (2021– )   

                                         GENRE  RATING  \
309        \nAction, Drama, Sci-Fi                 7.1   
435                        \nDrama                 8.5   
576   \nAnimation, Family, Fantasy                 8.0   
589     \nAction, Horror, Thriller                 6.4   
990      \nAction, Crime, Thriller                 6.5   
...                                        ...     ...   
9991      \nDrama, History, Horror                 NaN   
9992      \nDrama, History, Horror                 NaN   
9994   \nAdventure, Drama, Fantasy                 NaN   
9997   \nAdventure, Drama, Fantasy                 NaN   
9998   \nAdventure, Drama, Fantasy                 NaN   

                                               ONE-LINE  \
309   \nIn a future where a failed climate-change ex...   
435   \nA retired MMA fighter turned trainer juggles...   
576   \nA prince cursed to spend his days as a hideo...   
589   \nNever take your mistress on an annual guys' ...   
990   \nMei, a young girl whose memory holds a price...   
...                                                 ...   
9991                                     \nAdd a Plot\n   
9992                                     \nAdd a Plot\n   
9994                                     \nAdd a Plot\n   
9997                                     \nAdd a Plot\n   
9998                                     \nAdd a Plot\n   

                                                  STARS    VOTES  RunTime  \
309   \n    Director:\nBong Joon Ho\n| \n    Stars:\...  335,946    126.0   
435   \n            \n    Stars:\nFrank Grillo, \nKi...   16,565     43.0   
576   \n    Directors:\nGary Trousdale, \nKirk Wise\...  425,384     84.0   
589   \n    Director:\nCoralie Fargeat\n| \n    Star...   39,034    108.0   
990   \n    Director:\nBoaz Yakin\n| \n    Stars:\nJ...  106,164     94.0   
...                                                 ...      ...      ...   
9991                   \n    Director:\nBaran bo Odar\n      NaN      NaN   
9992                   \n    Director:\nBaran bo Odar\n      NaN      NaN   
9994  \n            \n    Stars:\nMorgan Taylor Camp...      NaN      NaN   
9997  \n    Director:\nJovanka Vuckovic\n| \n    Sta...      NaN      NaN   
9998  \n    Director:\nJovanka Vuckovic\n| \n    Sta...      NaN      NaN   

         Gross  
309     $4.56M  
435        NaN  
576   $218.97M  
589     $0.10M  
990    $17.14M  
...        ...  
9991       NaN  
9992       NaN  
9994       NaN  
9997       NaN  
9998       NaN  

[3182 rows x 9 columns]

...y también tenemos 3182 películas y/o series repetidas, sin contar las columnas posteriores. Podemos asumir que cada fila se trata de una serie con distintos episodios, o una película rebooteada.

# Hemos hecho una busqueda exhaustiva de distintos problemas e incoherencias que presenta el DataFrame. Solucionaremos cada uno de estos problemas mediante limpieza de datos.

In [ ]:
df_movies_clean = df_movies.copy()

- Eliminamos las columnas que tengan cierta cantidad de valores *nulos*.

In [ ]:
# para este ej p=60%
p = 0.6
thresh = int((1-p)*df_movies_clean.shape[0])
df_movies_clean.dropna(axis=1, thresh=thresh, inplace=True)

df_movies_clean.head()

MOVIES         YEAR  \
0                        Blood Red Sky       (2021)   
1  Masters of the Universe: Revelation     (2021– )   
2                     The Walking Dead  (2010–2022)   
3                       Rick and Morty     (2013– )   
4                      Army of Thieves       (2021)   

                                        GENRE  RATING  \
0      \nAction, Horror, Thriller                 6.1   
1  \nAnimation, Action, Adventure                 5.0   
2       \nDrama, Horror, Thriller                 8.2   
3  \nAnimation, Adventure, Comedy                 9.2   
4         \nAction, Crime, Horror                 NaN   

                                            ONE-LINE  \
0  \nA woman with a mysterious illness is forced ...   
1  \nThe war for Eternia begins again in what may...   
2  \nSheriff Deputy Rick Grimes wakes up from a c...   
3  \nAn animated series that follows the exploits...   
4  \nA prequel, set before the events of Army of ...   

                                               STARS    VOTES  RunTime  
0  \n    Director:\nPeter Thorwarth\n| \n    Star...   21,062    121.0  
1  \n            \n    Stars:\nChris Wood, \nSara...   17,870     25.0  
2  \n            \n    Stars:\nAndrew Lincoln, \n...  885,805     44.0  
3  \n            \n    Stars:\nJustin Roiland, \n...  414,849     23.0  
4  \n    Director:\nMatthias Schweighöfer\n| \n  ...      NaN      NaN

- Eliminamos la columna 'ONE-LINE'

In [ ]:
df_movies_clean.drop(columns=['ONE-LINE'], inplace=True)
df_movies_clean

MOVIES         YEAR  \
0                           Blood Red Sky       (2021)   
1     Masters of the Universe: Revelation     (2021– )   
2                        The Walking Dead  (2010–2022)   
3                          Rick and Morty     (2013– )   
4                         Army of Thieves       (2021)   
...                                   ...          ...   
9994                       The Imperfects     (2021– )   
9995                               Arcane     (2021– )   
9996                    Heart of Invictus     (2022– )   
9997                       The Imperfects     (2021– )   
9998                       The Imperfects     (2021– )   

                                           GENRE  RATING  \
0         \nAction, Horror, Thriller                 6.1   
1     \nAnimation, Action, Adventure                 5.0   
2          \nDrama, Horror, Thriller                 8.2   
3     \nAnimation, Adventure, Comedy                 9.2   
4            \nAction, Crime, Horror                 NaN   
...                                          ...     ...   
9994     \nAdventure, Drama, Fantasy                 NaN   
9995  \nAnimation, Action, Adventure                 NaN   
9996            \nDocumentary, Sport                 NaN   
9997     \nAdventure, Drama, Fantasy                 NaN   
9998     \nAdventure, Drama, Fantasy                 NaN   

                                                  STARS    VOTES  RunTime  
0     \n    Director:\nPeter Thorwarth\n| \n    Star...   21,062    121.0  
1     \n            \n    Stars:\nChris Wood, \nSara...   17,870     25.0  
2     \n            \n    Stars:\nAndrew Lincoln, \n...  885,805     44.0  
3     \n            \n    Stars:\nJustin Roiland, \n...  414,849     23.0  
4     \n    Director:\nMatthias Schweighöfer\n| \n  ...      NaN      NaN  
...                                                 ...      ...      ...  
9994  \n            \n    Stars:\nMorgan Taylor Camp...      NaN      NaN  
9995                                                 \n      NaN      NaN  
9996  \n    Director:\nOrlando von Einsiedel\n| \n  ...      NaN      NaN  
9997  \n    Director:\nJovanka Vuckovic\n| \n    Sta...      NaN      NaN  
9998  \n    Director:\nJovanka Vuckovic\n| \n    Sta...      NaN      NaN  

[9999 rows x 7 columns]

- Modificamos el nombre de la columna *RunTime*, para que quede en mayúsculas:

In [ ]:
df_movies_clean.rename(columns={'RunTime' : 'RUNTIME'}, inplace=True)
df_movies_clean.head()

MOVIES         YEAR  \
0                        Blood Red Sky       (2021)   
1  Masters of the Universe: Revelation     (2021– )   
2                     The Walking Dead  (2010–2022)   
3                       Rick and Morty     (2013– )   
4                      Army of Thieves       (2021)   

                                        GENRE  RATING  \
0      \nAction, Horror, Thriller                 6.1   
1  \nAnimation, Action, Adventure                 5.0   
2       \nDrama, Horror, Thriller                 8.2   
3  \nAnimation, Adventure, Comedy                 9.2   
4         \nAction, Crime, Horror                 NaN   

                                               STARS    VOTES  RUNTIME  
0  \n    Director:\nPeter Thorwarth\n| \n    Star...   21,062    121.0  
1  \n            \n    Stars:\nChris Wood, \nSara...   17,870     25.0  
2  \n            \n    Stars:\nAndrew Lincoln, \n...  885,805     44.0  
3  \n            \n    Stars:\nJustin Roiland, \n...  414,849     23.0  
4  \n    Director:\nMatthias Schweighöfer\n| \n  ...      NaN      NaN

- Eliminamos las filas duplicadas

In [ ]:
print(f"Antes : {len(df_movies_clean)} filas")
df_movies_clean.drop_duplicates(inplace=True)
print(f"Después : {len(df_movies_clean)} filas")


Antes : 9999 filas
Después : 9478 filas


- Eliminamos los saltos de línea en *GENRE* y *STARS*, y las ',' en *VOTES*

In [ ]:
df_movies_clean['GENRE'] = df_movies_clean['GENRE'].str.replace('\n', '')
df_movies_clean['STARS'] = df_movies_clean['STARS'].str.replace('\n', '')
df_movies_clean['VOTES'] = df_movies_clean['VOTES'].str.replace(',', '')

df_movies_clean.head()

MOVIES         YEAR  \
0                        Blood Red Sky       (2021)   
1  Masters of the Universe: Revelation     (2021– )   
2                     The Walking Dead  (2010–2022)   
3                       Rick and Morty     (2013– )   
4                      Army of Thieves       (2021)   

                                      GENRE  RATING  \
0      Action, Horror, Thriller                 6.1   
1  Animation, Action, Adventure                 5.0   
2       Drama, Horror, Thriller                 8.2   
3  Animation, Adventure, Comedy                 9.2   
4         Action, Crime, Horror                 NaN   

                                               STARS   VOTES  RUNTIME  
0      Director:Peter Thorwarth|     Stars:Peri B...   21062    121.0  
1                  Stars:Chris Wood, Sarah Michel...   17870     25.0  
2                  Stars:Andrew Lincoln, Norman R...  885805     44.0  
3                  Stars:Justin Roiland, Chris Pa...  414849     23.0  
4      Director:Matthias Schweighöfer|     Stars:...     NaN      NaN

- Llenemos los datos vacío según la columna

In [ ]:
values = {'YEAR' : 'NO INFORMATION', 'GENRE' : 'NO INFORMATION', 'RATING' : 0, 'STARS' : 'NO INFORMATION', 'VOTES' : 0, 'RUNTIME' : 0}
df_movies_clean.fillna(value=values, inplace=True)
df_movies_clean.head()

MOVIES         YEAR  \
0                        Blood Red Sky       (2021)   
1  Masters of the Universe: Revelation     (2021– )   
2                     The Walking Dead  (2010–2022)   
3                       Rick and Morty     (2013– )   
4                      Army of Thieves       (2021)   

                                      GENRE  RATING  \
0      Action, Horror, Thriller                 6.1   
1  Animation, Action, Adventure                 5.0   
2       Drama, Horror, Thriller                 8.2   
3  Animation, Adventure, Comedy                 9.2   
4         Action, Crime, Horror                 0.0   

                                               STARS   VOTES  RUNTIME  
0      Director:Peter Thorwarth|     Stars:Peri B...   21062    121.0  
1                  Stars:Chris Wood, Sarah Michel...   17870     25.0  
2                  Stars:Andrew Lincoln, Norman R...  885805     44.0  
3                  Stars:Justin Roiland, Chris Pa...  414849     23.0  
4      Director:Matthias Schweighöfer|     Stars:...       0      0.0

- Transformamos los datos de la columna *VOTES*, de *object* a *int64*

In [ ]:
df_movies_clean['VOTES'] = df_movies_clean['VOTES'].astype('int64')

df_movies_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9478 entries, 0 to 9998
Data columns (total 7 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   MOVIES   9478 non-null   object 
 1   YEAR     9478 non-null   object 
 2   GENRE    9478 non-null   object 
 3   RATING   9478 non-null   float64
 4   STARS    9478 non-null   object 
 5   VOTES    9478 non-null   int64  
 6   RUNTIME  9478 non-null   float64
dtypes: float64(2), int64(1), object(4)
memory usage: 592.4+ KB


In [ ]:
df_movies_clean

MOVIES         YEAR  \
0                           Blood Red Sky       (2021)   
1     Masters of the Universe: Revelation     (2021– )   
2                        The Walking Dead  (2010–2022)   
3                          Rick and Morty     (2013– )   
4                         Army of Thieves       (2021)   
...                                   ...          ...   
9993                            Totenfrau     (2022– )   
9995                               Arcane     (2021– )   
9996                    Heart of Invictus     (2022– )   
9997                       The Imperfects     (2021– )   
9998                       The Imperfects     (2021– )   

                                         GENRE  RATING  \
0         Action, Horror, Thriller                 6.1   
1     Animation, Action, Adventure                 5.0   
2          Drama, Horror, Thriller                 8.2   
3     Animation, Adventure, Comedy                 9.2   
4            Action, Crime, Horror                 0.0   
...                                        ...     ...   
9993               Drama, Thriller                 0.0   
9995  Animation, Action, Adventure                 0.0   
9996            Documentary, Sport                 0.0   
9997     Adventure, Drama, Fantasy                 0.0   
9998     Adventure, Drama, Fantasy                 0.0   

                                                  STARS   VOTES  RUNTIME  
0         Director:Peter Thorwarth|     Stars:Peri B...   21062    121.0  
1                     Stars:Chris Wood, Sarah Michel...   17870     25.0  
2                     Stars:Andrew Lincoln, Norman R...  885805     44.0  
3                     Stars:Justin Roiland, Chris Pa...  414849     23.0  
4         Director:Matthias Schweighöfer|     Stars:...       0      0.0  
...                                                 ...     ...      ...  
9993      Director:Nicolai Rohde|     Stars:Felix Kl...       0      0.0  
9995                                                          0      0.0  
9996      Director:Orlando von Einsiedel|     Star:P...       0      0.0  
9997      Director:Jovanka Vuckovic|     Stars:Morga...       0      0.0  
9998      Director:Jovanka Vuckovic|     Stars:Morga...       0      0.0  

[9478 rows x 7 columns]

Ya hemos limpiado el DataFrame. Se puede seguir limpiando y transformando de acuerdo a otros parámetros (separar los años, agrupar capítulos, separar los generos, etc), pero lo mantendremos así para que sea fácil de seguir la explicación.


## ¡Analicemos los datos!


### Veamos el promedio de Rating de las peliculas que tiene en su nombre Dead

In [ ]:
df_movies_clean[df_movies_clean['MOVIES'].str.contains('Dead')]['RATING'].mean()

6.2340425531914905

Esto sera mayor o menor a la media global?

In [ ]:
df_movies_clean['RATING'].mean()

5.951529858619962

Efectivamente podemos decir que las peliculas con la palabra dead en su nombre tiene un rating mayor a de la media.

In [ ]:
df1=df_movies_clean[df_movies_clean['MOVIES'].str.contains('Dead')]

### Veamos en que peliculas participa Angelina Jolie

In [ ]:
df_movies_clean[df_movies_clean['STARS'].str.contains('Angelina Jolie')]

MOVIES        YEAR  \
465               Kung Fu Panda 3      (2016)   
615                   The Tourist  (I) (2010)   
1457  First They Killed My Father      (2017)   

                                         GENRE  RATING  \
465   Animation, Action, Adventure                 7.1   
615               Action, Thriller                 6.0   
1457     Biography, Drama, History                 7.2   

                                                  STARS   VOTES  RUNTIME  
465       Directors:Alessandro Carloni, Jennifer Yuh...  141657     95.0  
615       Director:Florian Henckel von Donnersmarck|...  230565    103.0  
1457      Director:Angelina Jolie|     Stars:Sareum ...   16871    136.0

In [ ]:
df_movies_clean[df_movies_clean['STARS'].str.contains('Angelina Jolie')]['RATING'].mean()

6.766666666666667